### PR TITLE
WIP updating to use TLens v3

### PR DESCRIPTION
**Pull request by chanind**
This PR attempts to migrate to use the Huggingface bridge to allow running Gemma3 and more modern models backed by Huggingface. There are some remaining issues to be worked out still:

- Gemma-2 is erroring due to a TLens issue
- The GPT2 tests pass (`test_attribution_clt`) using relaxed tolerances - is this acceptable? I'm struggling to understand if this is just numerical issues or signs of something more serious.

### CODE REVIEW

**Comment by hannamw** [comment]
Is there a list somewhere of the TL v3 hook names / what parts of the model they correspond to? I'm happy to figure out what the corresponding hook names are, if there's some sort of documentation

**Comment by bryce13950** [comment]
@hannamw There is a doc in this [PR](https://github.com/TransformerLensOrg/TransformerLens/pull/1000/files) that breaks down how hooks correspond to new ones. There is also already an option to enable a compatibility layer by calling `enable_compatibility_mode`, which turns on all of the old hooks and shows you what they correspond to in the new structure.

**Comment by hannamw** [comment]
Thanks, that's exactly what I was looking for! Re: what these are supposed to do, the intent was to capture:
- the input to the MLP / Unembed (after the input layernorm)
- the output of the MLP (before the output layernorm, if there is post-layernorm)

That is, we wanted the inputs / outputs of these modules alone. But there didn't use to be any hooks for those, so we added these wrappers.

So, for `ReplacementUnembed`, I think these new hooks should do what we needed the wrapper for:
    - `hook_pre` corresponds to `unembed.hook_in`
    - `hook_post` corresponds to `unembed.hook_out`

However, for `ReplacementMLP`, it sounds like the new `mlp.hook_in` and `mlp.hook_out` hooks correspond to the old `hook_mlp_in` and `hook_mlp_out`, which are located before / after input / output layernorms respectively, right? So I'm not sure there are any TLv3 hookpoints that correspond to the ones we need at the moment.

@bryce13950 does this sound accurate? If there is a place where I could see the mappings from TLv3 hookpoints to TLv2 / HF model locations, I could be a bit more certain (especially about the unembed hooks).

**Comment by bryce13950** [comment]
If there is a model that does not have something configured, take a look at the architecture adapter. There may be a component mapping missing that we have not gotten to yet.

**Comment by hannamw** [comment]
Gotcha, thanks! In that case, I think that the `ReplacementMLP`'s `hook_in` / `hook_out` map directly to `mlp.hook_in` / `mlp.hook_out`. I think my confusion stemmed from the part of the documentation saying:

### MLP
- `blocks.{i}.mlp.hook_in`: (batch, pos, d_model)
  - *Legacy alias: `blocks.{i}.hook_mlp_in`*
- `blocks.{i}.mlp.hook_pre`: (batch, pos, d_mlp)
  - *Legacy alias: `blocks.{i}.hook_mlp_in` (via `mlp.in.hook_out`)*
- `blocks.{i}.mlp.hook_out`: (batch, pos, d_model)
  - *Legacy alias: `blocks.{i}.hook_mlp_out`*

I thought this meant that the new `blocks.{i}.mlp.hook_in` was equivalent to the current `blocks.{i}.hook_mlp_in`, but the latter is located before the MLP input layernorm (rather than immediately before the MLP).

**Comment by bryce13950** [comment]
@hannamw Thanks for bringing that up! I will bookmark this as a reminder before I close out that PR. I still have to verify all of the mappings one by one before I close it out, so it is very helpful to have that one already double checked!

File: tests/test_replacement_model.py [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+    assert torch.allclose(bridge_ctx.encoder_vecs, legacy_ctx.encoder_vecs, atol=1e-3, rtol=1e-3)
+    assert torch.allclose(
+        bridge_ctx.encoder_to_decoder_map, legacy_ctx.encoder_to_decoder_map, atol=1e-3, rtol=1e-3
+    )
+    assert torch.allclose(
+        bridge_ctx.decoder_locations, legacy_ctx.decoder_locations, atol=1e-3, rtol=1e-3
+    )
+
+
+def test_TransformerBridge_backward_gradients_differ_from_HookedTransformer():
```
[the remark this answers is not in the document]
Comment (chanind) [reply tests/test_replacement_model.py]: This tests passes for me

File: circuit_tracer/replacement_model.py [hunk 2]
```diff
@@ -185,43 +175,71 @@ def stop_gradient(acts, hook):
...
+        block = self.blocks[layer]
         cached = {}
 
         def cache_activations(acts, hook):
             cached["acts"] = acts
+            # Clear the cached result from previous forward pass
+            cached.pop("result", None)
 
         def add_skip_connection(acts: torch.Tensor, hook: HookPoint, grad_hook: HookPoint):
+            # There's a TLens bug that results in hooks getting called multiple times in compatibility mode.
```
Comment (hannamw) [inline circuit_tracer/replacement_model.py]: As mentioned in our call, it would be nice to revert this hack and make sure things still work, if the TransformerBridge issues are solved

File: circuit_tracer/replacement_model.py [hunk 3]
```diff
@@ -230,6 +248,10 @@ def add_skip_connection(acts: torch.Tensor, hook: HookPoint, grad_hook: HookPoin
             partial(add_skip_connection, grad_hook=subblock.hook_out_grad),
             is_permanent=True,
         )
+        # this is super hacky, but we need this in the hook registry or hooks won't be picked up
+        name = f"blocks.{layer}.{hook_out_str}.hook_out_grad"
```
Comment (hannamw) [inline circuit_tracer/replacement_model.py]: In the old TL implementation, this is handled by a `self.setup()` call that uses the `HookedRootModule.setup()` method to automatically set up the hook registry based on the `HookPoint` that are present. I'm not sure if such a thing exists in TransformerBridge, but that could be a solution.

File: circuit_tracer/replacement_model.py [hunk 4]
```diff
@@ -638,17 +685,21 @@ def intervention_hook(activations, hook, layer: int):
...
                 partial(calculate_delta_hook, layer=layer, layer_interventions=layer_interventions),
             )
             for layer, layer_interventions in interventions_by_layer.items()
         ]
 
         intervention_range = constrained_layers if constrained_layers else range(self.cfg.n_layers)
         intervention_hooks = [
-            (f"blocks.{layer}.{self.feature_output_hook}", partial(intervention_hook, layer=layer))
+            (
+                f"blocks.{layer}.{self.original_feature_output_hook}",
```
Comment (hannamw) [inline circuit_tracer/replacement_model.py]: Having looked over the entirety of this code, it seems like Claude changed the locations of the delta and intervention hooks to `self.original_feature_output_hook`, from `self.feature_output_hook`. It also added the former hook to the list of hooks to be frozen in the case of `constrained_layers`, but didn't remove the latter. 

This seems sort of weird to me, though it would only really matter for `constrained_layers` interventions, where we actually do any freezing of the MLP outputs. I would check the order in which things are being executed in TransformerBridge, but in general, the desired behavior for interventions with constrained layers is for each:
1. Freeze the MLP outputs to their original values. Compute the deltas corresponding to the interventions a that layer (if there are any).
2. Add the deltas into that layer's MLP outputs.

Anyway, you want to be sure that the model is not freezing the layer again after adding in the deltas, which could happen if feature_output_hook fires after original_feature_output_hook (again, unsure how this happens in TransformerBridge); if this doesn't happen, the change is harmless, I guess. But, I don't think that this change was necessary, and would revert it.

File: circuit_tracer/replacement_model.py [hunk 5]
```diff
@@ -185,43 +175,71 @@ def stop_gradient(acts, hook):
                 block.ln1_post.hook_scale.add_hook(stop_gradient, is_permanent=True)  # type: ignore
             if hasattr(block, "ln2_post"):
                 block.ln2_post.hook_scale.add_hook(stop_gradient, is_permanent=True)  # type: ignore
-            self.ln_final.hook_scale.add_hook(stop_gradient, is_permanent=True)  # type: ignore
-
-        for param in self.parameters():
+        self.ln_final.hook_scale.add_hook(stop_gradient, is_permanent=True)  # type: ignore
+
+        for name, param in self.named_parameters():
```
[the remark this answers is not in the document]
Comment (chanind) [reply circuit_tracer/replacement_model.py]: This is very strange and I can't explain it, but if I change this to `if param.is_leaf: param.requires_grad = False`, then `test_attribution_clt.py` fails on the small model (the full gpt2 model still passes though). 🤷‍♂️

File: circuit_tracer/replacement_model.py (hunk 3 above)
Comment (chanind) [reply circuit_tracer/replacement_model.py]: It looks like this isn't available in the bridge, but I found another method, `self._scan_existing_hooks(self, "")`, that seems to do this. Hopefully there's no side-effects to calling this method as it's an un-documented internal method 🤞

File: circuit_tracer/replacement_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hannamw) [reply circuit_tracer/replacement_model.py]: In my experience, when setting `requires_grad=False` on a tensor causes attribution to fail, it's often because that tensor was the leaf node that caused many other downstream tensors to require_grad; removing the grad requirement means that the backwards pass simply doesn't proceed down the path towards that tensor, since nothing on it requires a gradient anyway. Sometimes that means that the backwards pass no longer reaches a hook that it was supposed to reach, causing incorrect attributions. This is why we set `requires_grad=True` for the output of the embeds using a hook.

In this case, I'm not sure how the attention would cause this though. The gradients should just be flowing through the attention's OV circuit anyway - the embeds are upstream of everything, which should always cause the gradients to flow as needed. I'll think about this some more

File: circuit_tracer/replacement_model.py (hunk 4 above)
Comment (chanind) [reply circuit_tracer/replacement_model.py]: I tried reverting this change but find that `test_bridge_vs_legacy_clt_attribution_on_gpt2` no longer passes if I do 🤔
